### ISSUE TITLE
Idea: Pharmaverse general guidance for new contributors

### ISSUE THREAD

**Issue by rossfarrugia** [pharmaverse/pharmaverse#284]
**Describe your idea**
@ddsjoberg raised an idea to have some general pharmaverse guidance for brand new devs - e.g. resources to help learn R pkg dev, git/github etc - not specific to any package

I raised to the council and they liked the idea and we felt this could get added to this page: https://pharmaverse.org/contribute/ic/

If anyone has any useful resources then please share here and I could try and draft something, e.g. things like https://r-pkgs.org/ or if we can find any nice free git training

**Comment by rossfarrugia** [comment]
We're also thinking worth to publish a pharmaverse blog on this topic to raise awareness once the page is updated

**Comment by manciniedoardo** [comment]
This is a bit vague, but a bit of general (maybe obvious?) advice I would add to [this page](https://pharmaverse.org/contribute/ic/) is to consider starting out in a small niche in which you are interested. This could mean limiting yourself to just a single package in the pharamaverse, or even only one of the open issues in that package. The pharmaverse has grown quite large and it's a bit overwhelming at the start, but not every IC has to have their fingers in many pies.

**Comment by DamianRodziewicz** [comment]
We had an interesting discussion yesterday at the R Validation Hub community meeting. Very often, great developers are afraid and discouraged from contributing to Open Source for two reasons (among potential others):

1. Psychological - they don't want their code to be judged publicly, and they are afraid to take the first steps as these steps seem huge. I've been there as a programmer, and I know a lot of programmers who have also been there.
2. Technical - they don't know how to contribute. They may not know Git, and they may not understand the flow.

My suggestion that solves (1) and (2) is to create a step-by-step guide to 'your first OS contribution to pharmaverse' - where every person who would like to start contributing has to simply write one new unit test to any function. That could be a perfect ice breaker and can unblock contributors as they will see that it is not that difficult and scary to start adding value. By taking small steps, they can build up courage and confidence, and learn how fulfilling the whole initiative is. I have met people who struggled for years to push their first public commit, and once they finally did it they became very active contributors.

I suggest writing out all the steps a developer would have to take to write a test, for example, this one - https://github.com/pharmaverse/admiral/blob/main/tests/testthat/test-admiral_options.R#L12C1-L16C3 . Or maybe we could create an artificial issue to show the whole flow end-to-end where there is a specific issue we can reference in the whole process, how and where to ask the questions, how to create a pull request for a particular issue, etc. This could be even done in a short webinar where the presenter could share the screen and go through the whole process.

I am attaching some useful resources to learn from that I compiled a long time ago. I hope it helps!
[Useful resources to learn from.pdf](https://github.com/pharmaverse/pharmaverse/files/14359487/Useful.resources.to.learn.from.pdf)

**Comment by bms63** [comment]
I think the onboarding issue/video we made in admiral was incredibly helpful!!! The wall of documentation is just too much!!  Having a video of it is much more accessible.

Keeping it up to date is the tough part!!

**Comment by DamianRodziewicz** [comment]
@bms63 Ross sent me the link to the [admiral onboarding task and the video](https://github.com/pharmaverse/admiral/issues/1839) - I hadn't seen it before and it is precisely what I had in mind. So maybe one takeaway from my stream of thoughts could be that tests could be a good first step for contributors as they are not interfering with the package logic yet, but can already add value without any fear.

**Comment by cpiraux** [comment]
I recently made my first PR in an open-source package (xportr), and I would like to share my experience:

Last year, I began my involvement with the xportr R package, and I'd like to share my journey towards my first pull request (PR). As a statistical programmer mainly using SAS, with some knowledge of R, I initially had no experience with R package development. Nevertheless, I was eager to contribute to an open-source R package.

The first step in getting involved with an R package is to be aware that volunteers are needed. In my case, I noticed a call for volunteers posted on Slack for the xportr package. Since this package aligns with my expertise in metadata, I eagerly jumped at the opportunity.

With no prior experience in R package development and apprehensions about potentially breaking the code, I initially adopted an observational approach. I began by installing the package, running its functions, and then exploring the code to understand their implementation. I tested the functions to evaluate if they worked as expected. If I encountered any issues, I documented them in a GitHub issue.

After creating the GitHub issues with my findings, I began to follow the implementation process by another developer to become more familiar with GitHub. This involved observing how branches are created, commits are made, and comparisons between commits and branches are performed. I also learned about the functioning of comments, pull requests, and the review process.

Once his PR was completed, it was time for the review phase, during which I ran the new code to understand the changes and experiment with it. This step also involved two technical requirements: firstly, linking my GitHub account with RStudio to access the code in the branch, and secondly, understanding how to execute code within the branch. Since installing packages using 'install.packages' wouldn't function for development code, it was essential to know about the function 'devtools::load_all()' to execute the code.

As I became familiar with both the package and GitHub, I felt more confident to start my first implementation. I read the book 'R Packages' to gain a better understanding of R package development. The first chapter, entitled 'The Whole Game', provided a good overview of R package development and introduced useful functions such as 'load_all()', 'document()', 'test()', and 'check()'.

With this first implementation, I dived into the development process of the package, involving tasks such as creating a new branch, committing changes, writing a descriptive commit message (still not sure of the best approach), updating and adding new tests to cover the modifications, resolving merge conflicts, updating documentation, handling failing checks in CI/CD. Having a checklist in the PR template ([xportr-pull_request_template.md](https://github.com/atorus-research/xportr/blob/main/.github/pull_request_template.md)) was very helpful to keep all the necessary steps in mind.

The PR was submitted, and the reviewers provided useful feedback and valuable advice on my code. I implemented the comments, and the PR was approved and merged into the main branch.

After successfully merging my first PR, I feel more confident in contributing to open-source R packages. It's now time to select a new issue to tackle.

**Comment by rossfarrugia** [comment]
@cpiraux this is great! Would you mind if I include your story above in a pharmaverse blog (https://github.com/pharmaverse/blog/issues/134)? I'm going to be drafting this soon and I think hearing a real life story would inspire others. I could tag you in the review to check you're happy before we post.
